src/diagram/join.rs

```diff
@@ -0,0 +1,228 @@
+use std::any::TypeId;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use tracing::debug;
+
+use crate::{Builder, IterBufferable};
+
+use super::{DiagramError, DynOutput, NodeRegistry, SerializeMessage};
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct JoinOp {
+    pub(super) next: String,
+}
+
+pub(super) fn register_join_impl<T, Serializer>(registry: &mut NodeRegistry)
+where
+    T: Send + Sync + 'static,
+    Serializer: SerializeMessage<Vec<T>>,
+{
+    if registry.join_impls.contains_key(&TypeId::of::<T>()) {
+        return;
+    }
+
+    registry
+        .join_impls
+        .insert(TypeId::of::<T>(), Box::new(join_impl::<T>));
+
+    // FIXME(koonpeng): join_vec results in a SmallVec<[T; N]>, we can't serialize it because
+    // it doesn't implement JsonSchema, and we can't impl it because of orphan rule. We would need
+    // to create our own trait that covers `JsonSchema` and `Serialize`.
+    // register_serialize::<Vec<T>, Serializer>(registry);
+}
+
+fn join_impl<T>(builder: &mut Builder, outputs: Vec<DynOutput>) -> Result<DynOutput, DiagramError>
+where
+    T: Send + Sync + 'static,
+{
+    debug!("join outputs {:?}", outputs);
+
+    if outputs.is_empty() {
+        // do a empty join, in practice, this branch is never ran because [`WorkflowBuilder`]
+        // should error out if there is an empty join.
+        return Err(DiagramError::EmptyJoin);
+    }
+
+    let first_type = outputs[0].type_id;
+
+    let outputs = outputs
+        .into_iter()
+        .map(|o| {
+            // joins is only supported for outputs of the same type. This is because joins of
+            // different types produces a tuple and we cannot output a tuple as we don't
+            // know the number and order of join inputs at compile time.
+            // A workaround is to serialize them all the `serde_json::Value` or convert them to `Box<dyn Any>`.
+            // But the problem with `Box<dyn Any>` is that we can't convert it back to the original type,
+            // so nodes need to take a request of `JoinOutput<Box<dyn Any>>`.
```

Comment:
This might be too cumbersome for users in many cases.

What if we do:

```
{
    "type": "join",
    "next": "foo",
    "serialize": true
}
```

so users can decide within the diagram to send a serialized version of the data to the next node. I think it'll be much less troublesome for users to create nodes that support deserializing than nodes that can only connect to `JoinOutput<T>`.

Reply:
41a123b9e1d340ebfea021b82013237ce8ab095f

Reply:
What would you think of the `serialize` option being true by default? I suspect that's what users will want more often than not since the non-serialized version requires them to expect a specific size of `SmallVec`.

Reply:
I think it mostly depends on the type of app and the type of user. The end user (the user of the app that uses bevy_impulse diagram) would probably not care about perf, defaulting to serialize would make writing diagrams easier. The users of bevy_impulse (the devs of the app) might not want to end users to serialize, but perhaps defaulting to serialize is one way to push them to have their own "join" node specialized for the data they are handling.

Reply:
922db287189996ee529beb13bf38c4f901b01141